# Checks imports from higher layer into layer below (`project-fsd-architecture/layer-imports`)

<!-- end auto-generated rule header -->

Please describe the origin of the rule here.

## Rule Details

This rule aims to...

Examples of **incorrect** code for this rule:

```js

// fill me in

```

Examples of **correct** code for this rule:

```js

// fill me in

```

### Options

* `alias`: specify this option if you use alias.

* `ignoreImportPatterns`: specify regex to detect import from ignored files.

## When Not To Use It

Give a short description of when it would be appropriate to turn off this rule.

## Further Reading

If there are other links that describe the issue this rule addresses, please include them here in a bulleted list.
